### PR TITLE
GetInvocationCounter

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/307

### CODE REVIEW

File: neo/SmartContract/InteropService.cs [hunk 1]
```diff
@@ -499,6 +511,15 @@ private static bool Contract_Call(ApplicationEngine engine)
             if (currentManifest != null && !currentManifest.CanCall(contract.Manifest, method.GetString()))
                 return false;
 
+            if (engine.InvocationCounter.TryGetValue(contract.ScriptHash, out var counter))
+            {
+                engine.InvocationCounter[contract.ScriptHash] = counter + 1;
+            }
+            else
+            {
+                engine.InvocationCounter[contract.ScriptHash] = 1;
```
Comment (vncoelho) [inline neo/SmartContract/InteropService.cs]: Shargon, could this Dictionary be already initialized in 1 and avoid this if?

File: neo.UnitTests/UT_Syscalls.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            // Check the results
+
+            CollectionAssert.AreEqual
+                (
+                engine.ResultStack.Select(u => (int)((VM.Types.Integer)u).GetBigInteger()).ToArray(),
+                new int[]
+                {
+                    1, /* C */
+                    2, /* B */
+                    1, /* A */
```
Comment (vncoelho) [inline neo.UnitTests/UT_Syscalls.cs]: This should be 2 and the next should be 1, no?

Comment (vncoelho) [reply neo.UnitTests/UT_Syscalls.cs]: 1, /* C */
1, /* B */
2, /* A */
1, /* A */

File: neo/SmartContract/InteropService.cs (hunk 1 above)
Comment (shargon) [reply neo/SmartContract/InteropService.cs]: we need to increment the value, so you need to query the previous value

File: neo.UnitTests/UT_Syscalls.cs (hunk 2 above)
Comment (shargon) [reply neo.UnitTests/UT_Syscalls.cs]: check the issue, we call twice B

File: neo/SmartContract/InteropService.cs (hunk 1 above)
Comment (vncoelho) [reply neo/SmartContract/InteropService.cs]: Yes, but I mean, for the first time initialize already with 1 and then just increment.

Comment (vncoelho) [reply neo/SmartContract/InteropService.cs]: This would simplify the if.

File: neo.UnitTests/UT_Syscalls.cs (hunk 2 above)
Comment (vncoelho) [reply neo.UnitTests/UT_Syscalls.cs]: image

Comment (vncoelho) [reply neo.UnitTests/UT_Syscalls.cs]: A twice, no?

**Comment by igormcoelho** [comment]
@shargon can you take a look at https://github.com/neo-project/neo/issues/814 ? I think its related.

**Comment by shargon** [comment]
Is not related, is according to this issue https://github.com/neo-project/neo/issues/307

File: neo/SmartContract/InteropService.cs [hunk 3]
```diff
@@ -212,6 +213,17 @@ private static bool Runtime_Serialize(ApplicationEngine engine)
             return true;
         }
 
+        private static bool Runtime_GetInvocationCounter(ApplicationEngine engine)
+        {
+            if (!engine.InvocationCounter.TryGetValue(engine.CurrentScriptHash, out var counter))
+            {
+                counter = 1;
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]: I don't think this can happen.

File: neo/SmartContract/InteropService.cs [hunk 4]
```diff
@@ -499,6 +511,15 @@ private static bool Contract_Call(ApplicationEngine engine)
             if (currentManifest != null && !currentManifest.CanCall(contract.Manifest, method.GetString()))
                 return false;
 
+            if (engine.InvocationCounter.TryGetValue(contract.ScriptHash, out var counter))
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]: https://github.com/neo-project/neo/blob/e497d95c1aa446c319cd7dbae1b674c882c052cd/neo/SmartContract/ApplicationEngine.cs#L54-L57

I think we can move the counter logic into `ApplicationEngine_ContextLoaded()`.

File: neo/SmartContract/InteropService.cs (hunk 3 above)
Comment (shargon) [reply neo/SmartContract/InteropService.cs]: Yes, I couldn't happen, return false?
```suggestion
                return false;
```

File: neo.UnitTests/UT_Syscalls.cs (hunk 2 above)
Comment (shargon) [reply neo.UnitTests/UT_Syscalls.cs]: Fixed

File: neo/SmartContract/InteropService.cs (hunk 1 above)
Comment (shargon) [reply neo/SmartContract/InteropService.cs]: Initialize to 1 when it is not found

**Comment by shargon** [comment]
@erikzhang could you compute the gas cost?

**Comment by shargon** [comment]
The mock sometimes works, and other don't ...

File: neo/SmartContract/ApplicationEngine.cs [hunk 5]
```diff
@@ -19,6 +19,7 @@ public partial class ApplicationEngine : ExecutionEngine
         private readonly List<NotifyEventArgs> notifications = new List<NotifyEventArgs>();
         private readonly List<IDisposable> disposables = new List<IDisposable>();
 
+        public Dictionary<UInt160, int> InvocationCounter { get; } = new Dictionary<UInt160, int>();
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]:
```suggestion
        public IReadOnlyDictionary<UInt160, int> InvocationCounter { get; } = new Dictionary<UInt160, int>();
```

File: neo/SmartContract/InteropService.cs [hunk 6]
```diff
@@ -39,6 +39,7 @@ public static partial class InteropService
         public static readonly uint System_Runtime_GetTime = Register("System.Runtime.GetTime", Runtime_GetTime, 0_00000250);
         public static readonly uint System_Runtime_Serialize = Register("System.Runtime.Serialize", Runtime_Serialize, 0_00100000);
         public static readonly uint System_Runtime_Deserialize = Register("System.Runtime.Deserialize", Runtime_Deserialize, 0_00500000);
+        public static readonly uint System_Runtime_GetInvocationCounter = Register("System.Runtime.GetInvocationCounter", Runtime_GetInvocationCounter, 10000/*TODO Compute Gas*/);
```
Comment (erikzhang) [inline neo/SmartContract/InteropService.cs]:
```suggestion
        public static readonly uint System_Runtime_GetInvocationCounter = Register("System.Runtime.GetInvocationCounter", Runtime_GetInvocationCounter, 0_00000400);
```

File: neo/SmartContract/ApplicationEngine.cs (hunk 5 above)
Comment (shargon) [reply neo/SmartContract/ApplicationEngine.cs]: We need to be able to change the dictionary

**Pull request merged**